File: Days.34/Python/2.Construct-BST-from-Preorder-Traversal.py
# Construct BST from Preorder Traversal

class Node:
    def __init__(self,val):
        self.val=val
        self.left=None
        self.right=None

def BuildBST(indx,n,preorder,mini,maxi):
    if indx[0]>=n:
        return None
    if preorder[indx[0]]<mini or preorder[indx[0]]>maxi:
        return None
    root=Node(preorder[indx[0]])
    indx[0]+=1
    root.left=BuildBST(indx,n,preorder,mini,root.val)
    root.right=BuildBST(indx,n,preorder,root.val,maxi)
    return root

def Inorder(root):
    if root==None:
        return
    Inorder(root.left)
    print(root.val,end=" ")
    Inorder(root.right)

def main():
    n=int(input())
    preorder=list(map(int,input().split()))
    indx=[0]
    root=BuildBST(indx,n,preorder,-2**63,2**63)
    Inorder(root)
    print()




main()

File: Days.34/C++/2.Construct-BST-from-Preorder-Traversal.cpp
// Construct BST from Preorder Traversal

#include<iostream>
#include<climits>
#include<vector>
using namespace std;

class Node
{
    public:
    int val;
    Node *left;
    Node *right;
    Node(int val)
    {
        this->val=val;
        this->left=this->right=nullptr;
    }
};

Node* BuildBST(int &indx,int n,vector<int> &preorder,int mini,int maxi)
{
    if(indx>=n)
    {
        return nullptr;
    }
    if(preorder[indx]<mini or preorder[indx]>maxi)
    {
        return nullptr;
    }
    Node *root=new Node(preorder[indx]);
    indx++;
    root->left=BuildBST(indx,n,preorder,mini,root->val);
    root->right=BuildBST(indx,n,preorder,root->val,maxi);
    return root;
}

void Inorder(Node *root)
{
    if(root==nullptr)
    {
        return;
    }
    Inorder(root->left);
    cout<<root->val<< " ";
    Inorder(root->right);
}

int main()
{
    int n;
    cin>>n;
    vector<int> preorder(n);
    for(int i=0; i<n; i++)
    {
        cin>>preorder[i];
    }
    int indx=0;
    Node *root=BuildBST(indx,n,preorder,INT_MIN,INT_MAX);
    Inorder(root);
    cout<<"\n";
    return 0;
}